persistent-postgresql/Database/Persist/Postgresql/Util.hs

```diff
@@ -0,0 +1,30 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Database.Persist.Postgresql.Util (
+  mkPostgresUpdateText
+) where
+
+import Data.Monoid ((<>))
+import Data.Text (Text)
+import qualified Data.Text as T
+
+import Database.Persist (FieldNameDB, PersistEntity (..), PersistUpdate(..), Update(..), updateUpdate, fieldDB, FieldDef)
+
+
+mkPostgresUpdateText :: PersistEntity record => (FieldNameDB -> Text) -> (Text -> Text) -> Update record -> Text
+mkPostgresUpdateText escapeName refColumn x =
+  case updateUpdate x of
+    Assign -> n <> "=?"
+    Add -> T.concat [n, "=EXCLUDED.", refColumn n, "+?"]
+    Subtract -> T.concat [n, "=EXCLUDED.", refColumn n, "-?"]
+    Multiply -> T.concat [n, "=EXCLUDED.", refColumn n, "*?"]
+    Divide -> T.concat [n, "=EXCLUDED.", refColumn n, "/?"]
+    BackendSpecificUpdate up ->
+      error . T.unpack $ "mkUpdateText: BackendSpecificUpdate " <> up <> " not supported"
+  where
+    n = escapeName . fieldDB . updateFieldDef $ x
```

Comment:
Following on with the above - I think it'd be better to write this whole function like:

```haskell
mkPostgresUpdateText escapeName refColumn upd = 
    case upd of
        Update f op _ ->
            let 
                n = escapeName . fieldDB . persistField $ f
            in 
                case op of
                    Assign -> Right $  n <> "=?"
            -- ...
        BackendSpecificUpdate up ->
            Left $ errorMessage
```

That's more a style thing, though.